tfsdk/attribute.go

```diff
@@ -275,142 +300,117 @@ func (a Attribute) validate(ctx context.Context, req ValidateAttributeRequest, r
 		validator.Validate(ctx, req, resp)
 	}
 
-	if a.Attributes != nil {
-		nm := a.Attributes.GetNestingMode()
-		switch nm {
-		case NestingModeList:
-			l, ok := req.AttributeConfig.(types.List)
-
-			if !ok {
-				err := fmt.Errorf("unknown attribute value type (%T) for nesting mode (%T) at path: %s", req.AttributeConfig, nm, req.AttributePath)
-				resp.Diagnostics.AddAttributeError(
-					req.AttributePath,
-					"Attribute Validation Error",
-					"Attribute validation cannot walk schema. Report this to the provider developer:\n\n"+err.Error(),
-				)
+	a.validateAttributes(ctx, req, resp)
 
-				return
-			}
+	if a.DeprecationMessage != "" && attributeConfig != nil {
+		tfValue, err := attributeConfig.ToTerraformValue(ctx)
 
-			for idx := range l.Elems {
-				for nestedName, nestedAttr := range a.Attributes.GetAttributes() {
-					nestedAttrReq := ValidateAttributeRequest{
-						AttributePath: req.AttributePath.WithElementKeyInt(idx).WithAttributeName(nestedName),
-						Config:        req.Config,
-					}
-					nestedAttrResp := &ValidateAttributeResponse{
-						Diagnostics: resp.Diagnostics,
-					}
+		if err != nil {
+			resp.Diagnostics.AddAttributeError(
+				req.AttributePath,
+				"Attribute Validation Error",
+				"Attribute validation cannot convert value. Report this to the provider developer:\n\n"+err.Error(),
+			)
 
-					nestedAttr.validate(ctx, nestedAttrReq, nestedAttrResp)
+			return
+		}
 
-					resp.Diagnostics = nestedAttrResp.Diagnostics
-				}
-			}
-		case NestingModeSet:
-			s, ok := req.AttributeConfig.(types.Set)
+		if tfValue != nil {
+			resp.Diagnostics.AddAttributeWarning(
+				req.AttributePath,
+				"Attribute Deprecated",
+				a.DeprecationMessage,
+			)
+		}
+	}
+}
 
-			if !ok {
-				err := fmt.Errorf("unknown attribute value type (%T) for nesting mode (%T) at path: %s", req.AttributeConfig, nm, req.AttributePath)
-				resp.Diagnostics.AddAttributeError(
-					req.AttributePath,
-					"Attribute Validation Error",
-					"Attribute validation cannot walk schema. Report this to the provider developer:\n\n"+err.Error(),
-				)
+// validateAttributes performs all nested Attributes validation.
+func (a Attribute) validateAttributes(ctx context.Context, req ValidateAttributeRequest, resp *ValidateAttributeResponse) {
+	if !a.definesAttributes() {
+		return
+	}
 
-				return
-			}
+	nm := a.Attributes.GetNestingMode()
+	switch nm {
+	case NestingModeList:
+		l, ok := req.AttributeConfig.(types.List)
 
-			for _, value := range s.Elems {
-				tfValueRaw, err := value.ToTerraformValue(ctx)
+		if !ok {
+			err := fmt.Errorf("unknown attribute value type (%T) for nesting mode (%T) at path: %s", req.AttributeConfig, nm, req.AttributePath)
```

Comment:
I know you're not the one creating this message, but because we're here anyways, do we want to use the `attr.Type.String` method here instead of `%T`, to honor the provider developer's choices about how their type is represented?

Reply:
That sounds reasonable to me!